machine-runner/src/test-utils/mock-runner.ts

```diff
@@ -0,0 +1,303 @@
+      )
+    }
+    if (then) {
+      return then(snapshot)
+    }
+    return snapshot
+  }
+  // Below: public
+
+  const delay: Self['test']['delay'] = {
+    toggle: delayer.toggle,
+  }
+
+  const machine = createMachineRunnerInternal(
+    sub.subscribe,
+    async (events) => {
+      persisted.push(...events)
+      const pair = delayer.make()
+      const retval = pair[0].then(() => {
+        feed(events, { caughtUp: true })
+        return events.map((_) => mockMeta())
+      })
+      return retval
+    },
+    factory,
+    payload,
+  )
+
+  // TODO: make shareable with runner.ts
+  feed([], { caughtUp: true })
+
+  return {
+    ...machine,
+    test: {
+      assertAs,
+      feed,
+      delay,
+    },
+  }
+}
+
+// Utilities
+// ===============
+
+const mockMeta = () => ({
+  isLocalEvent: true,
+  tags: [],
+  timestampMicros: 0,
+  timestampAsDate: () => new Date(),
+  lamport: 1,
+  eventId: 'id1',
+  appId: 'test',
+  stream: 'stream1',
+  offset: 3,
+})
+
+// TODO: reuse in unit test
+
+type PromiseDelay = ReturnType<typeof PromiseDelay['make']>
+namespace PromiseDelay {
+  type Pair = [
+    Promise<void>,
+    { resolve: () => void; reject: () => void; isFinished: () => boolean },
+  ]
+
+  export const make = () => {
+    const data = {
+      isDelaying: false as boolean,
+      delayedCommands: [] as Pair[],
+    }
+
+    const make = (): Pair => {
+      let finished = false
+      const pair: Pair = [undefined as any, undefined as any]
+      pair[0] = new Promise<void>((resolve, reject) => {
+        pair[1] = {
+          resolve,
+          reject,
+          isFinished: () => finished,
+        }
+      }).finally(() => (finished = true))
+
+      if (data.isDelaying) {
+        data.delayedCommands.push(pair)
+      } else {
+        pair[1].resolve()
+      }
+      return pair
+    }
+
+    const toggle = async (
+      delayControl: { delaying: true } | { delaying: false; reject?: boolean },
+    ): Promise<void> => {
+      data.isDelaying = delayControl.delaying
+      if (delayControl.delaying) return
+      await Promise.all(
+        data.delayedCommands.map(([promise, control]) => {
+          if (delayControl.reject) {
+            control.reject()
+          } else {
+            control.resolve()
+          }
+          // eslint-disable-next-line @typescript-eslint/no-empty-function
+          return promise.catch(() => {})
+        }),
+      )
+      data.delayedCommands = []
+    }
+
+    return {
+      make,
+      toggle,
+    }
+  }
+}
+
+// TODO: reuse in unit test
+
+namespace Subscription {
+  export type CallbackFnOf<RegisteredEventsFactoriesTuple extends MachineEvent.Factory.Any[]> =
+    Parameters<SubscribeFn<RegisteredEventsFactoriesTuple>>[0]
+
+  export const make = <RegisteredEventsFactoriesTuple extends MachineEvent.Factory.Any[]>() => {
+    type Subscribe = SubscribeFn<RegisteredEventsFactoriesTuple>
+    type Cb = Parameters<Subscribe>[0]
```

Comment:
Nitpick - makes it dead obvious and easier to read the first time (IMO):
```suggestion
    type Callback = Parameters<Subscribe>[0]
```
If applied, don't forget to update the other places!

Reply:
Fixed on https://github.com/Actyx/machines/pull/30/commits/8038dd80eb7d588b8b5569da71377a3cfa29efb0